x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java

```diff
@@ -0,0 +1,376 @@
+import org.elasticsearch.node.Node;
+import org.elasticsearch.xpack.core.common.IteratingActionListener;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationServiceField;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationToken;
+import org.elasticsearch.xpack.core.security.authc.support.AuthenticationContextSerializer;
+import org.elasticsearch.xpack.core.security.support.Exceptions;
+import org.elasticsearch.xpack.core.security.user.AnonymousUser;
+import org.elasticsearch.xpack.core.security.user.SystemUser;
+import org.elasticsearch.xpack.core.security.user.User;
+import org.elasticsearch.xpack.security.operator.OperatorPrivileges.OperatorPrivilegesService;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+class AuthenticatorChain {
+
+    private static final Logger logger = LogManager.getLogger(AuthenticatorChain.class);
+
+    private final String nodeName;
+    private final boolean runAsEnabled;
+    private final OperatorPrivilegesService operatorPrivilegesService;
+    private final AnonymousUser anonymousUser;
+    private final boolean isAnonymousUserEnabled;
+    private final AuthenticationContextSerializer authenticationSerializer;
+    private final RealmsAuthenticator realmsAuthenticator;
+    private final List<Authenticator> allAuthenticators;
+    private final List<Authenticator> existingCredentialsAuthenticators;
+
+    AuthenticatorChain(
+        Settings settings,
+        OperatorPrivilegesService operatorPrivilegesService,
+        AnonymousUser anonymousUser,
+        AuthenticationContextSerializer authenticationSerializer,
+        ServiceAccountAuthenticator serviceAccountAuthenticator,
+        OAuth2TokenAuthenticator oAuth2TokenAuthenticator,
+        ApiKeyAuthenticator apiKeyAuthenticator,
+        RealmsAuthenticator realmsAuthenticator
+    ) {
+        this.nodeName = Node.NODE_NAME_SETTING.get(settings);
+        this.runAsEnabled = AuthenticationServiceField.RUN_AS_ENABLED.get(settings);
+        this.operatorPrivilegesService = operatorPrivilegesService;
+        this.anonymousUser = anonymousUser;
+        this.isAnonymousUserEnabled = AnonymousUser.isAnonymousEnabled(settings);
+        this.authenticationSerializer = authenticationSerializer;
+        this.realmsAuthenticator = realmsAuthenticator;
+        this.allAuthenticators = List.of(
+            serviceAccountAuthenticator,
+            oAuth2TokenAuthenticator,
+            apiKeyAuthenticator,
+            realmsAuthenticator
+        );
+        this.existingCredentialsAuthenticators = List.of(realmsAuthenticator);
+    }
+
+    void authenticateAsync(Authenticator.Context context, ActionListener<Authentication> listener) {
+        assert false == context.getDefaultOrderedRealmList().isEmpty() : "realm list must not be empty";
```

Comment:
Existing code has a branch to handle empty realm list. But the list can never to empty because the reserved realm is always there.

Reply:
After talking to Tim, this is only true for 8.0. In 7.x, the license could get downgraded in between calls which results in empty realm list.